src/System.Memory/src/System/SpanExtensions.cs

```diff
@@ -274,5 +274,116 @@ public static void CopyTo<T>(this T[] array, Span<T> destination)
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool Overlaps<T>(this Span<T> first, ReadOnlySpan<T> second, out int elementOffset)
+        {
+            return Overlaps((ReadOnlySpan<T>)first, second, out elementOffset);
+        }
+
+        /// <summary>
+        /// Determines whether two sequences overlap in memory.
+        /// </summary>
+        /// <remarks>
+        /// Two sequences overlap if they have positions in common and neither is empty.
+        /// Empty sequences do not overlap with any other sequence.
+        /// </remarks>
+        public static bool Overlaps<T>(this ReadOnlySpan<T> first, ReadOnlySpan<T> second)
+        {
+            if (first.IsEmpty || second.IsEmpty)
+            {
+                return false;
+            }
+
+            IntPtr byteOffset = Unsafe.ByteOffset(
+                ref first.DangerousGetPinnableReference(),
+                ref second.DangerousGetPinnableReference());
+
+            if (Unsafe.SizeOf<IntPtr>() == sizeof(int))
+            {
+                return (uint)byteOffset < (uint)(first.Length * Unsafe.SizeOf<T>()) ||
+                       (uint)byteOffset > (uint)-(second.Length * Unsafe.SizeOf<T>());
+            }
+            else
+            {
+                return (ulong)byteOffset < (ulong)((long)first.Length * Unsafe.SizeOf<T>()) ||
+                       (ulong)byteOffset > (ulong)-((long)second.Length * Unsafe.SizeOf<T>());
+            }
+        }
+
+        /// <summary>
+        /// Determines whether two sequences overlap in memory and outputs the element offset.
+        /// </summary>
+        /// <remarks>
+        /// Two sequences overlap if they have positions in common and neither is empty.
+        /// Empty sequences do not overlap with any other sequence. If the sequences
+        /// overlap, `elementOffset` is the number of elements by which the second sequence
+        /// is offset from the first sequence (i.e., second minus first). The number is
+        /// rounded away from zero if it is not a whole number, which can happen when
+        /// casting a sequence of a smaller type to a sequence of a larger type. If the
+        /// sequences do not overlap, `elementOffset` is meaningless and arbitrarily set to
+        /// zero.
+        /// </remarks>
+        public static bool Overlaps<T>(this ReadOnlySpan<T> first, ReadOnlySpan<T> second, out int elementOffset)
+        {
+            if (first.IsEmpty || second.IsEmpty)
+            {
+                elementOffset = 0;
+                return false;
+            }
+
+            IntPtr byteOffset = Unsafe.ByteOffset(
+                ref first.DangerousGetPinnableReference(),
+                ref second.DangerousGetPinnableReference());
+
+            if (Unsafe.SizeOf<IntPtr>() == sizeof(int))
+            {
+                if ((uint)byteOffset < (uint)(first.Length * Unsafe.SizeOf<T>()))
```

Comment:
Question: Can you please explain why the multiplication would never overflow? It it because of the 32-bit address space?

Otherwise, we could end up with condition being true/false based on the size of T.
```C#
int byteOffset = int.MaxValue;
int length = int.MaxValue - 1000000;
if ((uint)byteOffset < (uint)(length * 1)) // false
   Console.WriteLine("No overflow"); // Won't happen

if ((uint)byteOffset < (uint)(length * 2)) // true
   Console.WriteLine("Overflow");
```

Reply:
Yes. A span cannot refer to more memory than the size of the address space. So, on a 32-bit system, `first.Length * Unsafe.SizeOf<T>()` cannot be greater than 2<sup>32</sup>.

The multiplication can overflow an `int`, though. But that doesn't matter, because we're casting to `uint` immediately. Writing `(uint)first.Length * (uint)Unsafe.SizeOf<T>()` would be slightly more clear, but `(uint)(first.Length * Unsafe.SizeOf<T>())` gives the same result.

Reply:
`(uint)first.Length * (uint)Unsafe.SizeOf<T>()` might be better in case `checked` is switched on for the assembly at some point.

Maybe comment about 32-bit?

Reply:
Almost every line of the methods depends on unchecked arithmetic. Should I wrap each method body in an `unchecked` block?

What are you suggesting? 32-bit is not special; on a 64-bit system, `first.Length * Unsafe.SizeOf<T>()` cannot overflow as well because a span cannot refer to more than 2<sup>64</sup> bytes. The distinction between 32-bit and 64-bit is only made because there is no `nuint` type.

Reply:
Fair enough :)